### PR TITLE
Nonlinear eigenvalue solver interfaces  

### PR DESCRIPTION
**Pull request by fdkong**
The implementation of nonlinear eigenvalue solvers is in SLEPc, and we have the interface only at the moose side.  Keep moose simple by avoiding having logic algebra calculations  in moose.  

We already merged nonlinear inverse power into SLEPc, and the Newton-based should be ready soon.    Hopefully, all these solvers come up in the next release.

### CODE REVIEW

File: framework/src/utils/SlepcSupport.C [hunk 1]
```diff
@@ -231,6 +238,27 @@ setEigenSolverOptions(SolverParams & solver_params)
       Moose::PetscSupport::setSinglePetscOption("-eps_type", "jd");
       break;
 
+    case Moose::EST_NONLINEAR_POWER:
+#if !SLEPC_VERSION_LESS_THAN(3, 7, 3)
+      Moose::PetscSupport::setSinglePetscOption("-eps_type", "power");
+      Moose::PetscSupport::setSinglePetscOption("-st_type", "sinvert");
+      Moose::PetscSupport::setSinglePetscOption("-eps_power_nonlinear", "1");
+#else
+      mooseError("Nonlinear Inverse Power requires SLEPc 3.7.3 or higher  \n");
```
Comment (andrsd) [inline framework/src/utils/SlepcSupport.C]: We do not the trailing spaces and `\n`.

And below

File: framework/include/utils/SlepcSupport.h [hunk 2]
```diff
@@ -39,6 +39,11 @@ InputParameters getSlepcEigenProblemValidParams();
 void storeSlepcOptions(FEProblemBase & fe_problem, const InputParameters & params);
 void storeSlepcEigenProblemOptions(EigenProblem & eigen_problem, const InputParameters & params);
 void slepcSetOptions(FEProblemBase & problem);
+
+PetscErrorCode moose_slepc_eigen_formJacobianA(SNES snes, Vec x, Mat jac, Mat pc, void * ctx);
+PetscErrorCode moose_slepc_eigen_formJacobianB(SNES snes, Vec x, Mat jac, Mat pc, void * ctx);
+PetscErrorCode moose_slepc_eigen_formFunctionA(SNES snes, Vec x, Vec r, void * ctx);
+PetscErrorCode moose_slepc_eigen_formFunctionB(SNES snes, Vec x, Vec r, void * ctx);
```
Comment (andrsd) [inline framework/include/utils/SlepcSupport.h]: If this is our function, then `camelCaseStyle`.

File: python/TestHarness/testers/Tester.py [hunk 3]
```diff
@@ -31,6 +31,7 @@ def validParams():
         params.addParam('platform',      ['ALL'], "A list of platforms for which this test will run on. ('ALL', 'DARWIN', 'LINUX', 'SL', 'LION', 'ML')")
         params.addParam('compiler',      ['ALL'], "A list of compilers for which this test is valid on. ('ALL', 'GCC', 'INTEL', 'CLANG')")
         params.addParam('petsc_version', ['ALL'], "A list of petsc versions for which this test will run on, supports normal comparison operators ('<', '>', etc...)")
+        params.addParam('slepc_version', [],"A list of slepc versions for which this test will run on, supports normal comparison operators ('<', '>', etc...)")
```
Comment (andrsd) [inline python/TestHarness/testers/Tester.py]: Space after `,`

What is the reason for not having `'ALL'`?

File: python/TestHarness/testers/Tester.py [hunk 4]
```diff
@@ -269,7 +270,15 @@ def checkRunnableBase(self, options, checks, test_list=None):
...
             reasons['petsc_version'] = 'using PETSc ' + str(checks['petsc_version']) + ' REQ: ' + logic_reason + ' ' + petsc_version
 
-        # PETSc is being explicitly checked above
+        # Check for SLEPc versions
+        (slepc_status, logic_reason, slepc_version) = util.checkSlepcVersion(checks, self.specs)
+        if not slepc_status and len(self.specs['slepc_version']) != 0:
+            if slepc_version != None:
+                reasons['slepc_version'] = 'using SLEPc ' + str(checks['slepc_version']) + ' REQ: ' + logic_reason + ' ' + slepc_version
+            elif slepc_version == None:
+                reasons['slepc_version'] = 'SLEPc does not exist'
```
Comment (andrsd) [inline python/TestHarness/testers/Tester.py]: Did you mean `SLEPc is not installed`?

File: python/TestHarness/util.py [hunk 5]
```diff
@@ -248,6 +266,38 @@ def checkPetscVersion(checks, test):
             return (True, None, version)
     return (False, logic, version)
 
+
+# Break down slepc version logic in a new define
+def checkSlepcVersion(checks, test):
+    # User does not require anything
+    if len(test['slepc_version']) == 0:
+       return (False, None, None)
+    # These does not exist SLEPc
```
Comment (andrsd) [inline python/TestHarness/util.py]: I do not understand this comment

File: python/TestHarness/util.py [hunk 6]
```diff
@@ -248,6 +266,38 @@ def checkPetscVersion(checks, test):
...
+    if checks['slepc_version'] == None:
+       return (False, None, None)
+    # If any version of slepc works, return true immediately
```
Comment (andrsd) [inline python/TestHarness/util.py]: s/slepc/SLEPc/ and below

File: test/tests/problems/eigen_problem/ne.i [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+ ny = 8
+
+ uniform_refine = 0
+[]
+
+# the minimum eigenvalue of this problem is 2*(PI/a)^2;
+# Its inverse is 0.5*(a/PI)^2 = 5.0660591821169. Here a is equal to 10.
+
+[Variables]
+  active = 'u'
```
Comment (andrsd) [inline test/tests/problems/eigen_problem/ne.i]: Not needed

File: test/tests/problems/eigen_problem/ne.i [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+    use_displaced_mesh = true
+  [../]
+[]
+
+[Problem]
+  type = EigenProblem
+  eigen_problem_type = gen_non_hermitian
+  n_eigen_pairs = 1
+  n_basis_vectors = 18
+  which_eigen_pairs =  LARGEST_MAGNITUDE
```
Comment (andrsd) [inline test/tests/problems/eigen_problem/ne.i]: 1 space after `=`

File: test/tests/problems/eigen_problem/ne_deficient_b.i [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+    v = v
+    eigen_kernel = true
+    cof = -1.0
+  [../]
+  [./src_v]
+    type = CoupledForce
+    variable = v
+    v = u
+#    eigen_kernel = true
+#    cof = 1.0
```
Comment (andrsd) [inline test/tests/problems/eigen_problem/ne_deficient_b.i]: remove

File: test/tests/problems/eigen_problem/ne_deficient_b.i [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+  eigen_problem_type = gen_non_hermitian
+  n_eigen_pairs = 1
+  n_basis_vectors = 18
+  which_eigen_pairs = LARGEST_MAGNITUDE
+[]
+
+[Executioner]
+  type = Steady
+  eigen_solve_type = MONOLITH_NEWTON
+  petsc_options_iname = '-eps_power_snes_mf_operator '
```
Comment (andrsd) [inline test/tests/problems/eigen_problem/ne_deficient_b.i]: Remove that extra space inside `'...'`

File: test/tests/problems/eigen_problem/ne_coupled.i [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+  [../]
+[]
+
+[AuxKernels]
+  [./power_ak]
+    type = NormalizationAux
+    variable = power
+    source_variable = u
+    normalization = unorm
+# this coefficient will affect the eigenvalue.
```
Comment (andrsd) [inline test/tests/problems/eigen_problem/ne_coupled.i]: Indent the comment, this is not Fortran.

File: test/tests/problems/eigen_problem/tests [hunk 12]
```diff
@@ -73,4 +73,11 @@
     slepc_version = '>=3.7.4'
   [../]
 
+  [./coupled_system]
+    type = 'CSVDiff'
+    input = 'ne_coupled.i'
+    csvdiff = 'ne_coupled_eigenvalues_0001.csv'
+    slepc_version = '>=3.7.4'
+  [../]
+
```
Comment (andrsd) [inline test/tests/problems/eigen_problem/tests]: Extra empty line

**Review by andrsd — CHANGES_REQUESTED** [review]

**Comment by fdkong** [comment]
Moved to  #9343

**Pull request closed without merging**